### PR TITLE
add basic configuration to plugin

### LINKED ISSUE #332: Add basic configuration to plugin
MSFT want to be able to change at least the VM image the plugin chooses to use without revendoring the plugin.  Need a mechanism to allow this to be specified, either via the plugin reading a config file, or by accepting configuration at plugin instantiation time.

### PR DESCRIPTION
**Pull request by charlesakalugwu**
closes #332

### CODE REVIEW

**Comment by charlesakalugwu** [comment]
@jim-minter @kargakis @mjudeikis is this going in the right direction? We could in addition wire in the viper package to add support for plugin config overrides via a combination of config file, environment variables (e.g. prefixed by OPENSHIFT_AZURE_PLUGIN_) and command line flags.

File: pkg/plugin/config.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package plugin
+
+// PluginConfig contains dynamic plugin configuration
+type Config struct {
+	// The sync image to use
+	SyncImage string `json:"syncImage,omitempty"`
+
+	// The node image to use
+	NodeImage string `json:"nodeImage,omitempty"`
```
Comment (kargakis) [inline pkg/plugin/config.go]: I would suggest `vmImage` or `machineImage`. `nodeImage` is confusing even though we don't run kubelet inside a container (and there is also `origin-node`).

File: cmd/createorupdate/createorupdate.go [hunk 2]
```diff
@@ -26,10 +26,18 @@ import (
 
 var logLevel = flag.String("loglevel", "Debug", "valid values are Debug, Info, Warning, Error")
 
+var pc *plugin.Config
+
 // createOrUpdate simulates the RP
 func createOrUpdate(ctx context.Context, oc *v20180930preview.OpenShiftManagedCluster, entry *logrus.Entry) (*v20180930preview.OpenShiftManagedCluster, error) {
```
Comment (kargakis) [inline cmd/createorupdate/createorupdate.go]: Can you move plugin.Config to be passed as an argument in `createOrUpdate`? This is going to be useful once we share `createOrUpdate` between this code and the e2e framework.

**Comment by kargakis** [comment]
> We could in addition wire in the viper package to add support for plugin config overrides via a combination of config file, environment variables (e.g. prefixed by OPENSHIFT_AZURE_PLUGIN_) and command line flags.

Since the viper addition wouldn't actually be a part of the plugin code, I don't see any huge utility in bringing it in. Our e2e framework should reuse `createOrUpdate` so we are going to instantiate clusters in CI by passing pluginConfig. Where do you expect config overrides to be useful?

File: pkg/plugin/config.go (hunk 1 above)
Comment (charlesakalugwu) [reply pkg/plugin/config.go]: changed to `VmImage`

File: cmd/createorupdate/createorupdate.go (hunk 2 above)
Comment (charlesakalugwu) [reply cmd/createorupdate/createorupdate.go]: done

**Comment by charlesakalugwu** [comment]
> > We could in addition wire in the viper package to add support for plugin config overrides via a combination of config file, environment variables (e.g. prefixed by OPENSHIFT_AZURE_PLUGIN_) and command line flags.
> 
> Since the viper addition wouldn't actually be a part of the plugin code, I don't see any huge utility in bringing it in. Our e2e framework should reuse `createOrUpdate` so we are going to instantiate clusters in CI by passing pluginConfig. Where do you expect config overrides to be useful?

we could, for example, have a pipeline where all stages are configured with a default config file representing pluginConfig but having arbitrary individual stages able to have plugin.config.vmImage passed via flag as --vm-image and having the passed in value take precedence over the value in the default config file. Or being able to set `OPENSHIFT_AZURE_VM_IMAGE` from another stage...It was just a thought to enable super flexible 12-factor-style config in our project. It is however truly not necessary at this point.

**Comment by kargakis** [comment]
> but having arbitrary individual stages

Can you elaborate more with what you mean when you are saying "stages"? Are these stages during the install/scale/upgrade of a cluster or something else? The plugin config is supposed to hold configurables for a single request which means that during an install/scale/upgrade of a cluster, it's supposed to be the same all along the request (so the concept of "stages" does not make much sense in those scenarios)

**Comment by pweil-** [comment]
is the idea here that we may need to give MSFT the flexibility to change anything we're defaulting in the plugin but not exposing in the API?  

I see these options:

1.  expose things in the api
2.  create a new configuration input that overwrites the fields (multiple impl options here)
3.  have them set the environment variables already in use for images and see what creeps up next - aka procrastinate 

I'd suggest option 2 requires a discussion in our arch call before implementation.  In any case I think we should prioritize testing and SRE tooling over this right now unless it's blocking something.

**Comment by asalkeld** [comment]
also see https://github.com/openshift/openshift-azure/pull/492

**Comment by kargakis** [comment]
/hold

until this is discussed in the next arch call as per @pweil-'s request

**Comment by charlesakalugwu** [comment]
closing this and adding the work after #522 is merged since that is a basis for this work

**Pull request closed without merging**